### PR TITLE
New functions to close a mysql connection.

### PR DESCRIPTION
**Pull request by jarrocha**
This is a possible fix for issue #9415. Added a function in the mysql file and
a method for the metricset to close the db too.

I wanted to give it a shot since the last update was done more than 1 month ago. After researching the issue and feedback I thought about this possible solution. Is my approach good or do you prefer the metricset method doing all the work without calling to a function in the mysql file?

Let me know if there's a specific place where these functions should be called.

Also,
[https://github.com/elastic/beats/blob/824d443bc182a12f30859667da4739beef9c8606/metricbeat/mb/module/wrapper_test.go#L66](url)
is empty. Should it call these new functions for assertion? Just thinking of a possible new commit.

### CODE REVIEW

File: metricbeat/module/mysql/mysql.go [hunk 1]
```diff
@@ -106,3 +106,12 @@ func NewDB(dsn string) (*sql.DB, error) {
 	}
 	return db, nil
 }
+
+// CloseDB uses the database handle to close the connection and prevents future queries.
+func CloseDB(db *sql.DB) error {
+	err := db.Close()
```
Comment (andrewkroh) [inline metricbeat/module/mysql/mysql.go]: This function seems to be unnecessary.

File: metricbeat/module/mysql/status/status.go [hunk 2]
```diff
@@ -104,3 +104,13 @@ func (m *MetricSet) loadStatus(db *sql.DB) (map[string]string, error) {
 
 	return mysqlStatus, nil
 }
+
+// Close closes the database connection and prevents future queries.
+func (m *MetricSet) Close() error {
+	err := m.db.Close()
```
Comment (andrewkroh) [inline metricbeat/module/mysql/status/status.go]: This should have a nil check for the `db`. And `errors.Wrap()` can safely be called when `err` is nil and it just returns `nil`.

```
if m.db == nil {
  return nil
}
return errors.Wrap(m.db.Close(), "failed to close mysql database client")
```

Comment (andrewkroh) [reply metricbeat/module/mysql/status/status.go]: The same fix should be added to `metricbeat/module/mysql/galera_status/status.go`.

File: metricbeat/module/mysql/mysql.go (hunk 1 above)
Comment (jarrocha) [reply metricbeat/module/mysql/mysql.go]: Noted

File: metricbeat/module/mysql/status/status.go (hunk 2 above)
Comment (andrewkroh) [reply metricbeat/module/mysql/status/status.go]: No, this path isn't reachable with a nil `db`.

File: metricbeat/module/mysql/galera_status/status.go [hunk 3]
```diff
@@ -107,3 +107,11 @@ func (m *MetricSet) loadStatus(db *sql.DB) (map[string]string, error) {
 
 	return galeraStatus, nil
 }
+
+// Close closes the database connection and prevents future queries.
+func (m *MetricSet) Close() error {
+	if m.db == nil {
+		return errors.New("invalid database handle")
```
Comment (andrewkroh) [inline metricbeat/module/mysql/galera_status/status.go]: It should not be an error to have a nil `db`. This means that it never fetched or it never successfully initialized the DB client. In any case there's nothing for `Close()` to do, but it's not an error.

Comment (jarrocha) [reply metricbeat/module/mysql/galera_status/status.go]: Ok. I understand, I was missing the scope of work of this function. Now, I understand your initial code feedback. Thank you.

**Comment by andrewkroh** [comment]
jenkins, test this

**Comment by andrewkroh** [comment]
Can you please add an entry to the CHANGELOG.next.asciidoc file under the bugs / metricbeat section.

**Comment by andrewkroh** [comment]
jenkins, test this

**Comment by jarrocha** [comment]
Not a problem. But why does the automated test passes and the requested "jenkins, test this" fails?

**Comment by jarrocha** [comment]
Thank you for your patience.

**Comment by ruflin** [comment]
@jarrocha Thanks for this fix.

@jsoriano Should we also backport this?

**Comment by jsoriano** [comment]
Sure, doing it, thanks @jarrocha for fixing this!

**Comment by jarrocha** [comment]
Not a problem. I would like to keep going and grow. Do you have any suggestions? I have taken a look at this one which seems more challenging but you might know better? 
`https://github.com/elastic/beats/issues/6659`
Your feedback is appreciated. Thanks.

**Pull request merged**